pkg/graveler/committed/merge.go

```diff
@@ -0,0 +1,425 @@
+
+	case bytes.Equal(sourceRange.MinKey, destRange.MinKey) && bytes.Equal(sourceRange.MaxKey, destRange.MaxKey): // same bounds
+		baseRange, err := m.baseRangeGE(graveler.Key(sourceRange.MinKey))
+		if err != nil {
+			return err
+		}
+		if baseRange != nil && (sourceRange.ID == baseRange.ID || destRange.ID == baseRange.ID) {
+			if sourceRange.ID == baseRange.ID { // dest added changes
+				err = m.writeRange(destRange)
+			} else {
+				err = m.writeRange(sourceRange) // source added changes
+			}
+			if err != nil {
+				return err
+			}
+			m.haveSource = m.source.NextRange()
+			m.haveDest = m.dest.NextRange()
+		} else { // enter both ranges
+			m.haveSource = m.source.Next()
+			m.haveDest = m.dest.Next()
+		}
+
+	default: // ranges overlapping
+		m.haveSource = m.source.Next()
+		m.haveDest = m.dest.Next()
+	}
+	return nil
+}
+
+// applyBothKeys applies merging when both source and dest insides ranges
+func (m *merger) applyBothKeys(sourceValue *graveler.ValueRecord, destValue *graveler.ValueRecord) error {
+	c := bytes.Compare(sourceValue.Key, destValue.Key)
+	switch {
+	case c < 0: // source before dest
+		baseValue, err := m.baseKeyGE(sourceValue.Key)
+		if err != nil {
+			return err
+		}
+		if baseValue != nil && bytes.Equal(sourceValue.Identity, baseValue.Identity) { // dest deleted this record
+			m.haveSource = m.source.Next()
+		} else {
+			if baseValue != nil && bytes.Equal(sourceValue.Key, baseValue.Key) { // conflict
+				return graveler.ErrConflictFound
+			}
+			// source added this record
+			err := m.writeRecord(sourceValue)
+			if err != nil {
+				return fmt.Errorf("write source record: %w", err)
+			}
+			m.haveSource = m.source.Next()
+		}
+	case c > 0: // dest before source
+		baseValue, err := m.baseKeyGE(destValue.Key)
+		if err != nil {
+			return err
+		}
+		if baseValue != nil && bytes.Equal(destValue.Identity, baseValue.Identity) { // source deleted this record
+			m.haveDest = m.dest.Next()
+		} else {
+			if baseValue != nil && bytes.Equal(destValue.Key, baseValue.Key) { // conflict
+				return graveler.ErrConflictFound
+			}
+			// dest added this record
+			err := m.writeRecord(destValue)
+			if err != nil {
+				return fmt.Errorf("write dest record: %w", err)
+			}
+			m.haveDest = m.dest.Next()
+		}
+	default: // identical keys
+		baseValue, err := m.baseKeyGE(destValue.Key)
+		if err != nil {
+			return err
+		}
+		if !bytes.Equal(sourceValue.Identity, destValue.Identity) {
+			if baseValue != nil && (bytes.Equal(sourceValue.Identity, baseValue.Identity) ||
+				bytes.Equal(destValue.Identity, baseValue.Identity)) {
+				err := m.writeRecord(sourceValue)
+				if err != nil {
+					return fmt.Errorf("write record: %w", err)
+				}
+				m.haveSource = m.source.Next()
+				m.haveDest = m.dest.Next()
+				return nil
+			} else { // both changed the same key
+				return graveler.ErrConflictFound
+			}
+		}
+		// both added the same record
```

Comment:
Not precise, there is also the (more common) case of no change 
sourceValue.Identity == destValue.Identity == baseValue.Identity

Reply:
Done